dspy/clients/finetune.py

```diff
@@ -0,0 +1,133 @@
+from abc import abstractmethod
+from concurrent.futures import Future
+from enum import Enum
+import os
+from pathlib import Path
+from typing import List, Dict, Any, Optional
+import ujson
+
+from datasets.fingerprint import Hasher
+
+
+# TODO: Move to a centralized location with all the other environment variables
+# Set the directory to save the fine-tuned models
+def get_finetune_directory() -> str:
+    """Get the directory to save the fine-tuned models."""
+    # TODO: Move to a centralized location with all the other env variables
+    dspy_cachedir = os.environ.get("DSPY_CACHEDIR")
+    dspy_cachedir = dspy_cachedir or os.path.join(Path.home(), ".dspy_cache")
+    finetune_dir = os.path.join(dspy_cachedir, 'finetune')
+    finetune_dir = os.path.abspath(finetune_dir)
+    return finetune_dir
+
+
+FINETUNE_DIRECTORY = get_finetune_directory()
+
+
+class TrainingMethod(str, Enum):
+    """Enum class for training methods.
+    
+    When comparing enums, Python checks for object IDs, which means that the
+    enums can't be compared directly. Subclassing the Enum class along with the
+    str class allows for direct comparison of the enums.
+    """
+    SFT = "SFT"
+    Preference = "Preference"
+
+
+class TrainingStatus(str, Enum):
+    """Enum class for remote training status."""
+    not_started = "not_started"
+    pending = "pending"
+    running = "running"
+    succeeded = "succeeded"
+    failed = "failed"
+    cancelled = "cancelled"
+
+
+"""Dictionary mapping training methods to the data keys they require."""
+TRAINING_METHOD_TO_DATA_KEYS = {
```

Comment:
I am unsure how generic this can apply to different platforms. We can keep it with this PR, but may be changed in the future

Reply:
This is true -- our `TRAINING_METHOD` is actually a misnomer for `DATA_METHOD`. We will have the correct abstraction in the next PR...